acceptance/analyzer_test.go

```diff
@@ -0,0 +1,503 @@
+						filepath.Join("testdata", "analyzer", "cache-image"),
+					)
+					h.Run(t, cmd)
+				})
+
+				it.After(func() {
+					removeTestImage(t, cacheImageRepo)
+				})
+
+				it("ignores the cache", func() {
+					cmd := exec.Command(
+						"docker", "run", "--rm",
+						"--mount", "type=bind,source=/var/run/docker.sock,target=/var/run/docker.sock",
+						"--env", "CNB_REGISTRY_AUTH={}", // In practice, we never set this variable in the daemon case. Setting to avoid failure to stat docker config directory when initializing cache.
+						analyzeImage,
+						"/bin/bash", "-c",
+						fmt.Sprintf("%s -daemon -cache-image %s some-image; ls -alR /layers", analyzerPath, cacheImageRepo),
+					)
+					output := h.Run(t, cmd)
+
+					h.AssertNotMatch(t, output, "some-layer.sha")
+					h.AssertNotMatch(t, output, "some-layer.toml")
+				})
+			})
+
+			when("cache directory case", func() {
+				var (
+					cacheDir string
+					err      error
+				)
+
+				it.Before(func() {
+					cacheDir, err = ioutil.TempDir("", "analyzer-acceptance")
+					h.RecursiveCopy(t, filepath.Join("testdata", "analyzer", "cache-dir"), cacheDir)
+					// Some OSes (like macOS) use symlinks for the standard temp dir.
+					// Resolve it so it can be properly mounted by the Docker daemon.
+					cacheDir, err = filepath.EvalSymlinks(cacheDir)
+					h.AssertNil(t, err)
+				})
+
+				it.After(func() {
+					os.RemoveAll(cacheDir)
+				})
+
+				it("uses the provided cache", func() {
+					cmd := exec.Command(
+						"docker", "run", "--rm",
+						"--mount", "type=bind,source=/var/run/docker.sock,target=/var/run/docker.sock",
+						"--volume", fmt.Sprintf("%s:/cache", cacheDir),
+						analyzeImage,
+						"/bin/bash", "-c",
+						fmt.Sprintf("%s -daemon -cache-dir /cache some-image; ls -alR /layers", analyzerPath),
+					)
+					output := h.Run(t, cmd)
+
+					h.AssertMatch(t, output, "/layers/some-buildpack-id")
+					h.AssertMatch(t, output, "some-layer.sha")
+					h.AssertMatch(t, output, "some-layer.toml")
+				})
+
+				when("the provided cache directory isn't writeable", func() {
+					it("recursively chowns the directory", func() {
+						cmd := exec.Command(
+							"docker", "run", "--rm",
+							"--mount", "type=bind,source=/var/run/docker.sock,target=/var/run/docker.sock",
+							"--volume", fmt.Sprintf("%s:/cache", cacheDir),
+							analyzeImage,
+							"/bin/bash", "-c",
+							fmt.Sprintf("%s -daemon -cache-dir /cache some-image; ls -alR /cache", analyzerPath),
+						)
+						output := h.Run(t, cmd)
+
+						h.AssertMatch(t, output, "2222 3333 .+ \\.")
+						h.AssertMatch(t, output, "2222 3333 .+ committed")
+					})
+				})
+			})
+		})
+	})
+
+	when("registry case", func() {
+		it("writes analyzed.toml", func() {
+			cmd := exec.Command(
+				"docker", "run", "--rm",
+				"--env", "CNB_REGISTRY_AUTH={}",
+				analyzeImage,
+				"/bin/bash", "-c",
+				fmt.Sprintf("%s some-image; ls -al /layers", analyzerPath),
```

Comment:
I was talking to @yaelharel about how we would do validation of the outcomes and looked at this PR as reference. It throw us for a loop to identify how the analyzer was being executed. After the fact, I acknowledge that it's pretty clever but wonder if it's maybe to obscure for others as well. It may be it's solely on the format (aka use of `fmt.Sprintf`) and something like a function that makes it more apparent would help.

For example:
```suggestion
				analyzeAndExec(analyzerPath, "ls -al /layers", ...),
```

An alternative being a more explicit helper struct/function.

Reply:
Ooh. I like this - thanks for the suggestion :)

Reply:
I was playing around with this and I came up with the following: 
```
			cmd := exec.Command("docker", append([]string{
				"run", "--rm",
				"--env", "CNB_REGISTRY_AUTH={}",
				analyzeImage,
			}, analyzeAndExec(t, analyzerPath, "some-image", "ls -al /layers")...)...)
```

Not 100% sure if it's clearer... I could do away with passing in `t`.

Reply:
I am inclined to leave this as-is for now, given that the alternative implementation doesn't seem much clearer (at least, to me...). Also, some of the latest code has bash invocations *before* analyzer is run, such as `fmt.Sprintf("chown -R 9999:3333 /cache; chmod -R 775 /cache; %s -daemon -cache-dir /cache some-image; ls -alR /cache", analyzerPath)` and I don't see a helper function adding much improvement in those cases. Feel free to reopen the conversation @jromero @yaelharel if you disagree.

Reply:
I personally feel like this should be important given the loop it took us through. That said, it would always be improved after the fact. I think my main pain point is that at first glance it looks like it's bash (which it is) that is being executed instead of analyzer and therefore obfuscating the execution of analyzer as a whole. I'm sure with some more effort and thought a nice abstraction/utility could be built.

Reply:
@jromero I'll have another go at it. What do you think about the presentation above?

Reply:
I'm not a huge fan of wrapping the call to `analyzer` in bash. I've done it only where necessary to get the contents of the layers or cache directories after `analyzer` has run. As an alternative, I've considered letting the container run, then copying the contents of the directory that I care about to the host and inspecting it there (this is how I am checking the contents of analyzed.toml: https://github.com/buildpacks/lifecycle/pull/319/files#diff-72171b9647f1d25c895f3dcfa8e02b44R126); this approach is a bit more overhead though. @ekcasey what do you think?

Reply:
Looking at the code, it looks like keeping the container will be cleaner and more readable. My only concern is the performance but I guess that since the number of acceptance tests is not too high, the effect will be minor.

Reply:
Copying files out of the container is probably _slightly_ slower (fwiw `bash` take longer to launch so i wouldn't say this definitively without experimenting) and adds complexity to the test code. But, in the wild the `analyzer` won't be invoked with `bash` so the test is more realistic if we don't wrap with `bash`. With some good test helpers I think either approach could produce readable tests.

Also, in the near future the `analyzer` will support windows and we probably want to test the lifecycle in an environment w/o `bash`.

For the above reasons I lean towards copying the files out.

Reply:
I want to say that the latest changes should address this concern. Check out `h.DockerRun` and `h.DockerRunAndCopy`. Note also that some tests still use `WithBash()` - these are the ones that deal with Unix file permissions and will possibly be skipped on Windows. Thanks @jromero for the `WithFlags()` idea that I stole from pack :)

Reply:
Resolving this convo, but feel free to un-resolve or provide other feedback on the new approach.